packages/components/src/dialog/button-bar.tsx

```diff
@@ -0,0 +1,59 @@
+/**
+ * External dependencies
+ */
+import React, { isValidElement, cloneElement } from 'react';
+import type { ReactElement, ReactNode, FunctionComponent } from 'react';
+import classnames from 'classnames';
+
+export type BaseButton = {
+	action: string;
+	disabled?: boolean;
+	label: ReactNode;
+	className?: string;
+	additionalClassNames?: string;
+	isPrimary?: boolean;
+	onClick?: ( closeDialog: () => void ) => void;
+};
+
+export type Button = ReactElement | BaseButton;
+
+type Props = {
+	buttons: Button[];
+	baseClassName: string;
+	onButtonClick: ( button: BaseButton ) => void;
+};
+
+const ButtonBar: FunctionComponent< Props > = ( { buttons, baseClassName, onButtonClick } ) => {
+	if ( ! buttons ) {
+		return null;
+	}
+
+	return (
+		<div className={ baseClassName + '__action-buttons' }>
+			{ buttons.map( ( button, index ) => {
+				if ( isValidElement( button ) ) {
+					return cloneElement( button, { key: 'dialog-button-' + index } );
+				}
+
+				const classes = classnames( button.className || 'button', button.additionalClassNames, {
+					'is-primary': button.isPrimary || buttons.length === 1,
+				} );
+
+				return (
+					<button
+						key={ button.action }
```

Comment:
It would be better if all elements produced from the `button` array had a `key` from the same namespace, so to say. Instead of mixing `index`es and `actions`, let's just use `index` everywhere.

As the `buttons` list hardly ever changes, the `key` prop is just a formality we need to avoid a warning. Using `index` is fine.

Reply:
Sounds good to me!